froide/helper/breadcrumbs.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from typing import Generator, Optional, Sequence, Tuple, Union
+
+from django.template.context import Context
+from django.urls import NoReverseMatch, reverse
+from django.views import View
+
+logger = logging.getLogger(__name__)
+
+BreadcrumbTuple = Tuple[str, Union[str, None]]
+"""
+First item: link text
+Second item: link, or None
+"""
+
+BreadcrumbItems = Sequence[Union[str, BreadcrumbTuple]]
+
+
+class BreadcrumbView(View):
+    """
+    Your view must either provide a `breadcrumbs` attribute, or a
+    `get_breadcrumbs` method.
+
+    The `breadcrumbs` attribute is a sequence of strings or tuples,
+    e.g. ["Just text", ("A link", "url-pattern-name")].
+    The first element of the tuple specifies the link text, the second one
+    will be REVERSED to the link, so you can provide path names.
+
+    Using the `get_breadcrumbs` method, you can provide breadcrumbs as
+    above, with the difference that urls WILL NOT BE REVERSED.
+    You can also directly return a `Breadcrumbs` instance.
+
+    Example implementation:
+
+    class MyView(BreadcrumbView):
+        def get_breadcrumbs():
+            return [("Foo", "/foo/"), "Bar"]
+    """
+
+    breadcrumbs: Optional[Sequence]
+
+    def get_breadcrumbs(self) -> Union[BreadcrumbItems, Breadcrumbs]:
+        raise NotImplementedError("No breadcrumb provider implemented")
+
+
+@dataclass
+class BreadcrumbItem:
+    # will be displayed as the link text
+    title: str
```

Comment:
Is `str` the right type for (lazy) translatable strings? https://docs.djangoproject.com/en/5.0/topics/i18n/translation/#lazy-translations

Reply:
I think `BreadcrumbItem` shouldn't deal with Promises. The `Breadcrumbs` class calls `str` before passing the title to `BreadcrumbItem`